File: ueueo-framework/ueueo-object-extending/src/main/java/com/ueueo/data/objectextending/modularity/EntityExtensionConfiguration.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using JetBrains.Annotations;
using Volo.Abp.Localization;

namespace Volo.Abp.ObjectExtending.Modularity;

public class EntityExtensionConfiguration
{
    [NotNull]
    protected ExtensionPropertyConfigurationDictionary Properties;//  { get; }

    [NotNull]
    public List<Action<ObjectExtensionValidationContext>> Validators;//  { get; }

    public Dictionary<String, Object> Configuration;//  { get; }

    public EntityExtensionConfiguration()
    {
        Properties = new ExtensionPropertyConfigurationDictionary();
        Validators = new List<Action<ObjectExtensionValidationContext>>();
        Configuration = new Dictionary<String, Object>();
    }

    [NotNull]
    public   EntityExtensionConfiguration AddOrUpdateProperty<TProperty>(
        @Nonnull String propertyName,
        @Nullable Action<ExtensionPropertyConfiguration> configureAction = null)
    {
        return AddOrUpdateProperty(
            typeof(TProperty),
            propertyName,
            configureAction
        );
    }

    [NotNull]
    public   EntityExtensionConfiguration AddOrUpdateProperty(
        @Nonnull Type propertyType,
        @Nonnull String propertyName,
        @Nullable Action<ExtensionPropertyConfiguration> configureAction = null)
    {
        Objects.requireNonNull(propertyType, nameof(propertyType));
        Objects.requireNonNull(propertyName, nameof(propertyName));

        var propertyInfo = Properties.GetOrAdd(
            propertyName,
            () => new ExtensionPropertyConfiguration(this, propertyType, propertyName)
        );
        configureAction?.Invoke(propertyInfo);

        NormalizeProperty(propertyInfo);

        if (!propertyInfo.UI.Lookup.Url.IsNullOrEmpty())
        {
            AddLookupTextProperty(propertyInfo);
            propertyInfo.UI.OnTable.IsVisible = false;
        }
        return this;
    }

    private void AddLookupTextProperty(ExtensionPropertyConfiguration propertyInfo)
    {
        var lookupTextPropertyName = $"{propertyInfo.Name}_Text";
        var lookupTextPropertyInfo = Properties.GetOrAdd(
           lookupTextPropertyName,
           () => new ExtensionPropertyConfiguration(this, typeof(String), lookupTextPropertyName)
       );

        lookupTextPropertyInfo.DisplayName = propertyInfo.DisplayName;
    }

    [NotNull]
    public   ImmutableList<ExtensionPropertyConfiguration> GetProperties()
    {
        return Properties.Values.ToImmutableList();
    }

    private static void NormalizeProperty(ExtensionPropertyConfiguration propertyInfo)
    {
        if (!propertyInfo.Api.OnGet.IsAvailable)
        {
            propertyInfo.UI.OnTable.IsVisible = false;
        }

        if (!propertyInfo.Api.OnCreate.IsAvailable)
        {
            propertyInfo.UI.OnCreateForm.IsVisible = false;
        }

        if (!propertyInfo.Api.OnUpdate.IsAvailable)
        {
            propertyInfo.UI.OnEditForm.IsVisible = false;
        }
    }
}

File: ueueo-frameworks/ueueo-object-extending/src/main/java/com/ueueo/data/objectextending/modularity/EntityExtensionConfiguration.java
package com.ueueo.data.objectextending.modularity;

import com.ueueo.data.objectextending.ObjectExtensionValidationContext;
import org.apache.commons.lang3.StringUtils;
import org.springframework.lang.NonNull;
import org.springframework.lang.Nullable;

import java.util.*;
import java.util.function.Consumer;

public class EntityExtensionConfiguration {
    @NonNull
    protected Map<String, ExtensionPropertyConfiguration> properties;

    @NonNull
    private List<Consumer<ObjectExtensionValidationContext>> validators;

    private Map<String, Object> configuration;

    public EntityExtensionConfiguration() {
        properties = new HashMap<>();
        validators = new ArrayList<>();
        configuration = new HashMap<>();
    }

    @NonNull
    public EntityExtensionConfiguration addOrUpdateProperty(
            @NonNull Class<?> propertyType,
            @NonNull String propertyName,
            @Nullable Consumer<ExtensionPropertyConfiguration> configureAction) {
        Objects.requireNonNull(propertyType);
        Objects.requireNonNull(propertyName);

        ExtensionPropertyConfiguration propertyInfo = properties.computeIfAbsent(
                propertyName,
                key -> new ExtensionPropertyConfiguration(this, propertyType, propertyName)
        );
        if (configureAction != null) {
            configureAction.accept(propertyInfo);
        }

        normalizeProperty(propertyInfo);

        if (StringUtils.isNotBlank(propertyInfo.getUi().getLookup().getUrl())) {
            addLookupTextProperty(propertyInfo);
            propertyInfo.getUi().getOnTable().setVisible(false);
        }
        return this;
    }

    private void addLookupTextProperty(ExtensionPropertyConfiguration propertyInfo) {
        String lookupTextPropertyName = propertyInfo.getName() + "_Text";
        ExtensionPropertyConfiguration lookupTextPropertyInfo = properties.computeIfAbsent(
                lookupTextPropertyName,
                key -> new ExtensionPropertyConfiguration(this, String.class, lookupTextPropertyName)
        );

        lookupTextPropertyInfo.setDisplayName(propertyInfo.getDisplayName());
    }

    @NonNull
    public List<ExtensionPropertyConfiguration> getProperties() {
        return new ArrayList<>(properties.values());
    }

    private static void normalizeProperty(ExtensionPropertyConfiguration propertyInfo) {
        if (!propertyInfo.getApi().getOnGet().isAvailable()) {
            propertyInfo.getUi().getOnTable().setVisible(false);
        }

        if (!propertyInfo.getApi().getOnCreate().isAvailable()) {
            propertyInfo.getUi().getOnCreateForm().setVisible(false);
        }

        if (!propertyInfo.getApi().getOnUpdate().isAvailable()) {
            propertyInfo.getUi().getOnEditForm().setVisible(false);
        }
    }

    @NonNull
    public List<Consumer<ObjectExtensionValidationContext>> getValidators() {
        return validators;
    }

    public Map<String, Object> getConfiguration() {
        return configuration;
    }
}
